### PR TITLE
MOBT-389: Modification to the mosg__model_run underpinnings

### PR DESCRIPTION
**Pull request by bayliffe**
As it stands the `mosg__model_run` attribute is inconsistent with the `mosg__model_configuration` attribute. The former includes the name and cycle of every model that is passed into blending, regardless of whether this is zero-weighted in the final blend. The latter includes only the non-zero-weighted contributions.

This PR sought to address this mismatch such that our metadata is consistent and accurate. In the process it became apparent that adding the weights into the attribute, as long desired, was a relatively modest extension to the approach taken. As a result this PR addresses the mismatch and also adds in the model weight contributions to the `mosg__model_run` attribute.

Weather symbols and the summary symbols do not include weights as the quantities are not particularly useful, nor straightforward to make useful.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver_tests/acceptance/SHA256SUMS [hunk 1]
```diff
@@ -674,17 +674,17 @@ a699ad6697b5ffff8d0b7ef46c54ae46d545ba7141e30e351373b82cf998bb08  ./weighted_ble
...
 d08df3e0ecb68e7307fcf0085b263a7baecf4f811b3acee9f3a8f34616d13a03  ./weighted_blending/three_models/enukxhrly/20190101T0400Z-PT0004H00M-precip_rate.nc
-d197c8f5f1a08caddaa5efabac4a52d75185889d64d2c4b72cbba2f2c0f42a20  ./weighted_blending/three_models/kgo.nc
+a723b5cc5a2ba050a33dcc569ba67b93e8043d50bbd868cfcd065d000ae52b39  ./weighted_blending/three_models/kgo.nc
 30062d9c23527634a20cae6e7cf7a163aaee361b06ee800c8bea70a7a0f412ec  ./weighted_blending/three_models/nc/20190101T0400Z-PT0001H00M-precip_rate.nc
 850d48dfea1d1e5b696324f0e57a52fce3e3128008c58ccaca82e31598196c67  ./weighted_blending/three_models/ukvx/20190101T0400Z-PT0002H00M-precip_rate.nc
-63b4691e4d7e246b58e01c84f3c8cf82a6cdc7e14068ddd96418e8c0054e2bcf  ./weighted_blending/weights_from_dict/kgo.nc
+1eac7df64f49df69aabadc49c7ddaad7d2acd72dc775c99930a963bc6567cb03  ./weighted_blending/weights_from_dict/kgo.nc
 c2b71855f260544ca75aaa5900f3297b58cee1b3b64eec5805662e1927d5fdab  ./wet-bulb-temperature-integral/basic/input.nc
 651c80cd4ab7c720c59c6c0c2c4a2ade85aee81c2ab0c697b577ca87a36d57f2  ./wet-bulb-temperature-integral/basic/with_id_attr/kgo.nc
```
Comment (MoseleyS) [inline improver_tests/acceptance/SHA256SUMS:689]: Would there be any benefit to updating the inputs to the weather code tests? I'm inclined to say no.

**Review by MoseleyS — CHANGES_REQUESTED** [review]

File: improver/blending/__init__.py [hunk 2]
```diff
@@ -32,3 +32,5 @@
 
 MODEL_BLEND_COORD = "model_id"
 MODEL_NAME_COORD = "model_configuration"
+RECORD_COORD = "blend_record"
+WEIGHT_PRECISION = "1.3f"
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/blending/__init__.py:36]: Changed everywhere.

File: improver_tests/blending/test_utilities.py [hunk 3]
```diff
@@ -66,13 +74,58 @@ def cycle_cube_fixture():
...
+@pytest.fixture
+def cycle_cube_with_blend_record():
+    cubes = setup_cycle_cube()
+    updated_cubes = CubeList()
+    for cube in cubes.slices_over("forecast_reference_time"):
+        time = (
+            cube.coord("forecast_reference_time").cell(0).point.strftime("%Y%m%dT%H%MZ")
+        )
+
+        blend_record_coord = AuxCoord([f"uk_det:{time}:1.000"], long_name=RECORD_COORD)
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/blending/test_utilities.py:97]: Done.

File: improver_tests/blending/test_utilities.py [hunk 4]
```diff
@@ -66,13 +74,58 @@ def cycle_cube_fixture():
...
+        cube.add_aux_coord(blend_record_coord)
+        updated_cubes.append(cube)
+    return updated_cubes.merge_cube()
+
+
+def create_weights(cube, blending_coord, weights):
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/blending/test_utilities.py:103]: Changed.

File: improver_tests/blending/test_utilities.py [hunk 5]
```diff
@@ -66,13 +74,58 @@ def cycle_cube_fixture():
...
+        weights_cube.remove_coord(crd)
+    weights_cube.data = weights
+    weights_cube.rename("weights")
+    weights_cube.units = 1
+
+    return weights_cube
+
+
+@pytest.fixture
+def cycle_weights(weights):
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/blending/test_utilities.py:122]: This is just a fixture that returns a list of weights suitable for cycle blending. I'm not sure I want to encapsulate a full description of the process into the fixture name, so I've made it more explicit as "cycle_blending_weights". Everywhere the fixture is used this makes clear what is being passed in. I've added a doc-string to describe what it is doing.

File: improver_tests/blending/test_utilities.py [hunk 6]
```diff
@@ -158,116 +274,261 @@ def test_update_blended_metadata(model_cube, forecast_period):
...
+    )
+    record_run_attr = "mosg__model_run"
+    cubes = list(model_cube_with_blend_record.slices_over(MODEL_BLEND_COORD))
+
+    apply_record_run_attr(model_cube, cubes, record_run_attr)
+    assert model_cube.attributes[record_run_attr] == expected
+
+    cubes = CubeList(model_cube_with_blend_record.slices_over(MODEL_BLEND_COORD))
+    apply_record_run_attr(model_cube, cubes, record_run_attr)
+    assert model_cube.attributes[record_run_attr] == expected
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/blending/test_utilities.py:402]: Good idea, done.

File: improver_tests/blending/test_utilities.py [hunk 7]
```diff
@@ -158,116 +274,261 @@ def test_update_blended_metadata(model_cube, forecast_period):
...
+):
+    """Test that the apply record method can remove weights from the resulting
+    attribute if so desired. If this results in any duplicates (as weights were
+    the differentiating feature) these should be removed."""
+
+    record_run_attr = "mosg__model_run"
+    cube = model_cube_with_blend_record
+    cube.coord(RECORD_COORD).points = [
+        "uk_det:20171110T0000Z:0.750",
+        "uk_det:20171110T0000Z:0.250",
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/blending/test_utilities.py:436]: This is testing the situation with weather symbols, where multiple input cubes have the same model inputs (all potentially with the same weights in fact). These inputs are never merged into a single cube, hence the ability to pass in a cubelist, which makes this situation possible. Without removing duplicates, and with the use of the discard_weights option for weather symbols, we would end up with lots of duplicate entries of uk_det:20221020T1200Z\nuk_det:20221020T1200Z\n etc.

File: improver_tests/acceptance/SHA256SUMS (hunk 1 above)
Comment (bayliffe) [reply improver_tests/acceptance/SHA256SUMS:689]: I've modified the basic test inputs to include a weight. The KGO is unchanged, demonstrating that the weights are thrown away as expected.

**Review by MoseleyS — CHANGES_REQUESTED** [review]
There's a code style issue that needs resolving.

**Review by MoseleyS — APPROVED** [review]

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks @bayliffe 👍 

I'll continue to look at the unit tests but I've added some minor comments below.

File: improver/blending/calculate_weights_and_blend.py [hunk 8]
```diff
@@ -300,10 +315,14 @@ def process(
 
         coords_to_remove = get_coords_to_remove(cube, self.blend_coord)
 
+        weights = None
         if len(cube.coord(self.blend_coord).points) > 1:
             weights = self._calculate_blending_weights(cube)
             cube, weights = self._remove_zero_weighted_slices(cube, weights)
 
+        if record_run_attr is not None and weights is not None:
+            cube = update_record_run_weights(cube, weights, self.blend_coord)
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/blending/calculate_weights_and_blend.py:324]: I'll leave this then.

File: improver_tests/blending/test_utilities.py [hunk 9]
```diff
@@ -95,6 +163,56 @@ def model_cube_fixture():
...
+        "uk_ens:20171109T2300Z:{uk_ens_weight:{WEIGHT_FORMAT}}\n"
+        "uk_ens:20171110T0000Z:{uk_ens_weight:{WEIGHT_FORMAT}}\n"
+        "uk_ens:20171110T0100Z:{uk_ens_weight:{WEIGHT_FORMAT}}",
+    ]
+
+
+@pytest.fixture
+def model_blend_record() -> List[str]:
+    """Return a blend record template."""
+    return model_blend_record_template()
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver_tests/blending/test_utilities.py:188]: Hmm, not sure if calling a globally defined function is okay in pytests, but it works, so I've removed the fixture.

File: improver/blending/calculate_weights_and_blend.py [hunk 10]
```diff
@@ -300,10 +315,14 @@ def process(
 
         coords_to_remove = get_coords_to_remove(cube, self.blend_coord)
 
+        weights = None
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/blending/calculate_weights_and_blend.py:318]: Done.

File: improver/blending/utilities.py [hunk 11]
```diff
@@ -100,9 +107,20 @@ def get_coords_to_remove(cube: Cube, blend_coord: str) -> Optional[List[str]]:
...
+        except CoordinateNotFoundError:
+            crds_to_remove = []
+        else:
+            crds_to_remove = [RECORD_COORD]
+
         if blend_coord == MODEL_BLEND_COORD:
-            return [MODEL_BLEND_COORD, MODEL_NAME_COORD]
+            crds_to_remove.extend([MODEL_BLEND_COORD, MODEL_NAME_COORD])
+        if crds_to_remove:
+            return crds_to_remove
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/blending/utilities.py:123]: Changed. I slightly moved this around, so please check it still makes sense to you with the empty list just defined once.

File: improver/blending/utilities.py [hunk 12]
```diff
@@ -310,11 +340,112 @@ def set_record_run_attr(
...
+        blend_point = cslice.coord(blend_coord).cell(0).point
+        weight = weights.extract(
+            Constraint(coord_values={blend_coord: lambda cell: cell == blend_point})
+        )
+        run_records = cslice.coord(RECORD_COORD).points[0].split("\n")
+        updated_records = []
+        for run_record in run_records:
+            components = run_record.rsplit(":", 1)
+            key = components[0]
+            value = float(components[-1]) * weight.data
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/blending/utilities.py:446]: Changed the name of key to model_cycle.

File: improver/blending/utilities.py [hunk 13]
```diff
@@ -235,44 +253,58 @@ def _get_cycletime_point(cube: Cube, cycletime: str) -> int64:
     return round_close(cycletime_point, dtype=np.int64)
 
 
-def set_record_run_attr(
+def store_record_run_attr(
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/blending/utilities.py:256]: Adopted.

File: improver_tests/blending/test_utilities.py [hunk 14]
```diff
@@ -158,116 +298,249 @@ def test_update_blended_metadata(model_cube, forecast_period):
...
-      - cubes with unique attributes that are combined
-      - cubes with identical attributes attributes that are combined such that
-        only a single entry is returned."""
+    information on the cubes. The resulting cubes have an additional
+    auxiliary RECORD_COORD that stores the cycle and model information."""
 
     record_run_attr = "mosg__model_run"
     model_id_attr = "mosg__model_configuration"
-    cubes = [model_cube[index] for index in indices]
+    cubes = cubes = [model_cube[0], model_cube[1]]
```
Comment (gavinevans) [inline improver_tests/blending/test_utilities.py:308]: Should this be this?
```suggestion
    cubes = [model_cube[0], model_cube[1]]
```

**Review by gavinevans — CHANGES_REQUESTED** [review]

File: improver/blending/utilities.py [hunk 15]
```diff
@@ -310,11 +340,112 @@ def set_record_run_attr(
...
+        )
 
-    cycle_strings.sort()
-    for cube in cubelist:
-        cube.attributes[record_run_attr] = "\n".join(cycle_strings)
+        record_coord = AuxCoord([run_attr], long_name=RECORD_COORD)
+        cube.add_aux_coord(record_coord)
+
+
+def apply_record_run_attr(
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/blending/utilities.py]: Adopted record_run_coord_to_attr

File: improver/wxcode/weather_symbols.py [hunk 16]
```diff
@@ -564,6 +561,11 @@ def create_symbol_cube(self, cubes: Union[List[Cube], CubeList]) -> Cube:
             optional_attributes=optional_attributes,
             data=np.ma.masked_all_like(template_cube.data).astype(np.int32),
         )
+        if self.record_run_attr and self.model_id_attr is not None:
+            apply_record_run_attr(
+                symbols, set(cubes), self.record_run_attr, discard_weights=True
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/wxcode/weather_symbols.py:571]: Added.

File: improver_tests/blending/test_utilities.py (hunk 14 above)
Comment (bayliffe) [reply improver_tests/blending/test_utilities.py:302]: I really wanted to make sure that cubes was set. Changed.

**Review by gavinevans — APPROVED** [review]

**Pull request merged**